pkg/autoscaler/autoscaler_server.go

```diff
@@ -0,0 +1,134 @@
+
+	"github.com/golang/glog"
+)
+
+// AutoScaler determines the number of replicas to run
+type AutoScaler struct {
+	K8sClient    K8sClientInterface
+	Controller   controller.Interface
+	ConfigMapKey string
+	PollPeriod   time.Duration
+	Clock        clock.Clock
+	StopCh       chan struct{}
+	ReadyCh      chan<- struct{} // For testing.
+}
+
+func NewAutoScaler(c *options.AutoScalerConfig) (*AutoScaler, error) {
+	var controller controller.Interface
+	switch c.Mode {
+	case laddercontroller.ControllerType:
+		controller = laddercontroller.NewLadderController()
+	default:
+		return nil, fmt.Errorf("not a supported control mode: %v", c.Mode)
+	}
+	newK8sClient, err := NewKubernetesClient(c.Namespace, c.Target)
+	if err != nil {
+		return nil, err
+	}
+	return &AutoScaler{
+		K8sClient:    newK8sClient,
+		Controller:   controller,
+		ConfigMapKey: c.ConfigMap,
+		PollPeriod:   time.Second * time.Duration(c.PollPeriodSeconds),
+		Clock:        clock.RealClock{},
+		StopCh:       make(chan struct{}),
+		ReadyCh:      make(chan struct{}, 1),
+	}, nil
+}
+
+// Run periodically counts the number of nodes and cores, estimates the expected
+// number of replicas, compares them to the actual replicas, and
+// updates the target resource with the expected replicas if necessary.
+func (s *AutoScaler) Run() {
+	firstTimeCh := make(chan struct{}, 1)
+	firstTimeCh <- struct{}{}
+	ticker := s.Clock.Tick(s.PollPeriod)
+	s.ReadyCh <- struct{}{} // For testing.
+
+	for {
+		select {
+		case <-firstTimeCh:
+			s.pollAPIServer()
+		case <-ticker:
+			s.pollAPIServer()
+		case <-s.StopCh:
+			return
+		}
+	}
+}
+
+func (s *AutoScaler) pollAPIServer() {
+	// Query the apiserver for the cluster status --- number of nodes and cores
+	clusterStatus, err := s.K8sClient.GetClusterStatus()
+	if err != nil {
+		glog.Errorf("error while counting nodes: %v\n", err)
+		return
+	}
+	glog.V(4).Infof("Total nodes %5d, schedulable nodes: %5d\n", clusterStatus.totalNodes, clusterStatus.schedulableNodes)
+	glog.V(4).Infof("Total cores %5d, schedulable cores: %5d\n", clusterStatus.totalCores, clusterStatus.schedulableCores)
+
+	// Fetch autoscaler ConfigMap data from apiserver
+	configMap, err := (s.K8sClient.FetchConfigMap(s.K8sClient.GetNamespace(), s.ConfigMapKey))
+	if err != nil {
+		glog.Errorf("error fetching scaler params: %s", err)
+		return
+	}
+
+	// Query the autoscaler for the expected replicas number
+	expReplicas, err := s.Controller.GetExpectedNumOfReplicas(configMap.Data[laddercontroller.ControllerType], configMap.Version, clusterStatus.schedulableNodes, clusterStatus.schedulableCores)
+	if err != nil {
+		glog.Errorf("error while calculating expected replicas number: %v\n", err)
+		return
+	}
+	glog.V(4).Infof("Expected replica count: %3d\n", expReplicas)
+
+	targetScaler, err := s.K8sClient.GetScaler()
+	if err != nil {
+		glog.Errorf("error while getting targetScaler: %v\n", err)
+		return
+	}
+
+	// Query the apiserver for this pod's information.
```

Comment:
Is there a reason to query and then set rather than just setting the replicas every time?  Best case it is the same number of API calls, worst case it is more.